### DISCUSSION TITLE
Spawning language server with cmd: `rust-analyzer` failed

### DISCUSSION THREAD

**Discussion by longguzzz** [LunarVim/LunarVim#3396] [Q&A] [answered]
I have installed the server.
![image](https://user-images.githubusercontent.com/96779663/199393537-afedbed0-360a-4a60-a1b9-382e2ac62c49.png)
But lvim still prompts "Spawning language server with cmd: `rust-analyzer` failed. The language server is either not installed, missing from PATH, or not executable."
The rust-analyzer is in `nvim-data\mason\packages\rust-analyzer\`. Is it the right path?

I also tried adding the code below to `config.lua` and tried `:LvimCacheReset` but it doen't work.
```
local rust_opts = {
    capabilities = require("cmp_nvim_lsp").default_capabilities(vim.lsp.protocol.make_client_capabilities()),
    flags = {
      debounce_text_changes = 150,
    },
    on_attach = function(client, bufnr)
      client.resolved_capabilities.document_formatting = false
      client.resolved_capabilities.document_range_formatting = false
      local function buf_set_keymap(...)
        vim.api.nvim_buf_set_keymap(bufnr, ...)
      end
      require("keybindings").mapLSP(buf_set_keymap)
    end,
  }
require("lspconfig")["rust_analyzer"].setup(rust_opts)
```

I think the information in document is confusing. Where should I insert the code below in ?
![image](https://user-images.githubusercontent.com/96779663/199396062-e2be5b33-3f1c-4e60-a197-529a384bdbe0.png)

**Comment by longguzzz** [comment]
![image](https://user-images.githubusercontent.com/96779663/199401481-d981bf76-5b0e-4d4f-ba55-fc81b32a4ce3.png)
Golsp works.

![image](https://user-images.githubusercontent.com/96779663/199401520-d9f68ab9-ffd3-44b9-ab54-622cac564b54.png)
Rust-analyzer does not work. However, it seems know that "Other clients that match the filetype: rust" and that `rust-analyer` is an executable cmd.

No record in `nvim-data/lsp.log`. But I found something like "nvim.40516.0 terminfo_start:359: uv_tty_set_mode failed: invalid argument" in filie `nvim-data/log`.

**Comment by LostInTheLogs** [comment]
are you on latest master?

**Comment by LostInTheLogs** [comment, marked as answer]
if you want a more advanced rust setup use https://github.com/LunarVim/starter.lvim/tree/rust-ide config

**Reply by longguzzz** [reply]
I reinstall lvim with `Invoke-WebRequest https://raw.githubusercontent.com/LunarVim/LunarVim/master/utils/installer/install.ps1 -UseBasicParsing | Invoke-Expression`.
And I try your config as follow, I add different line in your config to default generated config. Then, I run `:PackerSync`.
```
-- Enable powershell as your default shell
vim.opt.shell = "pwsh.exe -NoLogo"
vim.opt.shellcmdflag =
  "-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;"
vim.cmd [[
		let &shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
		let &shellpipe = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
		set shellquote= shellxquote=
  ]]

-- Set a compatible clipboard manager
vim.g.clipboard = {
  copy = {
    ["+"] = "win32yank.exe -i --crlf",
    ["*"] = "win32yank.exe -i --crlf",
  },
  paste = {
    ["+"] = "win32yank.exe -o --lf",
    ["*"] = "win32yank.exe -o --lf",
  },
}

-- general
lvim.log.level = "warn"
lvim.format_on_save = true
lvim.colorscheme = "tokyonight"

lvim.leader = "space"

lvim.keys.normal_mode["<C-s>"] = ":w<cr>"

lvim.builtin.alpha.active = true
lvim.builtin.alpha.mode = "dashboard"
lvim.builtin.terminal.active = false

lvim.builtin.nvimtree.setup.diagnostics.enable = nil
lvim.builtin.nvimtree.setup.filters.custom = nil
lvim.builtin.nvimtree.setup.git.enable = nil
lvim.builtin.nvimtree.setup.update_cwd = nil
lvim.builtin.nvimtree.setup.update_focused_file.update_cwd = nil
lvim.builtin.nvimtree.setup.view.side = "left"
lvim.builtin.nvimtree.setup.renderer.highlight_git = nil
lvim.builtin.nvimtree.setup.renderer.icons.show.git = nil

lvim.builtin.treesitter.ensure_installed = {
  "c",
  "lua",
  "rust",
	"toml",
}
lvim.builtin.dap.active = true
lvim.builtin.treesitter.ignore_install = { "haskell" }
lvim.builtin.treesitter.highlight.enable = true

vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "rust_analyzer" })

local formatters = require "lvim.lsp.null-ls.formatters"
formatters.setup {
	{ command = "stylua", filetypes = { "lua" } },
}


local mason_path = vim.fn.glob(vim.fn.stdpath "data" .. "/mason/")
local codelldb_adapter = {
	type = "server",
	port = "${port}",
	executable = {
		command = mason_path .. "bin/codelldb",
		args = { "--port", "${port}" },
		-- On windows you may have to uncomment this:
		-- detached = false,
	},
}

pcall(function()
	require("rust-tools").setup {
		tools = {
			executor = require("rust-tools/executors").termopen, -- can be quickfix or termopen
			reload_workspace_from_cargo_toml = true,
			runnables = {
				use_telescope = true,
			},
			inlay_hints = {
				auto = true,
				only_current_line = false,
				show_parameter_hints = false,
				parameter_hints_prefix = "<-",
				other_hints_prefix = "=>",
				max_len_align = false,
				max_len_align_padding = 1,
				right_align = false,
				right_align_padding = 7,
				highlight = "Comment",
			},
			hover_actions = {
				border = "rounded",
			},
			on_initialized = function()
				vim.api.nvim_create_autocmd({ "BufWritePost", "BufEnter", "CursorHold", "InsertLeave" }, {
					pattern = { "*.rs" },
					callback = function()
						local _, _ = pcall(vim.lsp.codelens.refresh)
					end,
				})
			end,
		},
		dap = {
			adapter = codelldb_adapter,
		},
		server = {
			on_attach = function(client, bufnr)
				require("lvim.lsp").common_on_attach(client, bufnr)
				local rt = require "rust-tools"
				vim.keymap.set("n", "K", rt.hover_actions.hover_actions, { buffer = bufnr })
			end,

			capabilities = require("lvim.lsp").common_capabilities(),
			settings = {
				["rust-analyzer"] = {
					lens = {
						enable = true,
					},
					checkOnSave = {
						enable = true,
						command = "clippy",
					},
				},
			},
		},
	}
end)

-- CHANGED --
lvim.builtin.dap.on_config_done = function(dap)
	dap.adapters.codelldb = codelldb_adapter
	dap.configurations.rust = {
		{
			name = "Launch file",
			type = "codelldb",
			request = "launch",
			program = function()
				return vim.fn.input("Path to executable: ", vim.fn.getcwd() .. "/", "file")
			end,
			cwd = "${workspaceFolder}",
			stopOnEntry = false,
		},
	}
end

vim.api.nvim_set_keymap("n", "<m-d>", "<cmd>RustOpenExternalDocs<Cr>", { noremap = true, silent = true })

lvim.builtin.which_key.mappings["L"] = {
	name = "Rust",
	r = { "<cmd>RustRunnables<Cr>", "Runnables" },
	t = { "<cmd>lua _CARGO_TEST()<cr>", "Cargo Test" },
	m = { "<cmd>RustExpandMacro<Cr>", "Expand Macro" },
	c = { "<cmd>RustOpenCargo<Cr>", "Open Cargo" },
	p = { "<cmd>RustParentModule<Cr>", "Parent Module" },
	d = { "<cmd>RustDebuggables<Cr>", "Debuggables" },
	v = { "<cmd>RustViewCrateGraph<Cr>", "View Crate Graph" },
	R = {
		"<cmd>lua require('rust-tools/workspace_refresh')._reload_workspace_from_cargo_toml()<Cr>",
		"Reload Workspace",
	},
	o = { "<cmd>RustOpenExternalDocs<Cr>", "Open External Docs" },
}

lvim.plugins = {
	"simrat39/rust-tools.nvim",
	{
		"saecki/crates.nvim",
		tag = "v0.3.0",
		requires = { "nvim-lua/plenary.nvim" },
		config = function()
			require("crates").setup {
				null_ls = {
					enabled = true,
					name = "crates.nvim",
				},
			}
		end,
	},
	{
		"j-hui/fidget.nvim",
		config = function()
			require("fidget").setup()
		end,
	},
}
```

But the problem still exists.
![image](https://user-images.githubusercontent.com/96779663/199793838-37076165-3cb2-4077-b0dd-fa7f86c72dcb.png)
![image](https://user-images.githubusercontent.com/96779663/199793685-39b7c702-d4b6-4bb9-9c30-6e099a6f4d6d.png)
![image](https://user-images.githubusercontent.com/96779663/199793774-526e240c-1f51-4183-8259-73ecaf1ed1ea.png)


However, lua works well.
![image](https://user-images.githubusercontent.com/96779663/199795372-3fb12f58-53d9-4d8b-a5cd-32cdd48cc1a0.png)
![image](https://user-images.githubusercontent.com/96779663/199796010-da57c078-0125-47fa-948c-9992966c1928.png)

**Reply by LostInTheLogs** [reply]
what does `:lua print(vim.fn.executable("rust-analyzer"))` print?

**Reply by longguzzz** [reply]
> what does `:lua print(vim.fn.executable("rust-analyzer"))` print?

1

**Reply by LostInTheLogs** [reply]
are you on latest master?

**Reply by LostInTheLogs** [reply]
oh, you reinstalled so you are

**Reply by longguzzz** [reply]
I run `Invoke-WebRequest https://raw.githubusercontent.com/LunarVim/LunarVim/master/utils/installer/install.ps1 -UseBasicParsing | Invoke-Expression` to install. The version shows rolling-7d57046. I think it may be not the latest. But the command automatically choose this version.

**Reply by LostInTheLogs** [reply]
that's weird

**Reply by LostInTheLogs** [reply]
go to appdata/lunarvim/lvim and run git checkout master there

**Reply by longguzzz** [reply]
Yes, that's really weird... and consumes a lot of time... 
(I may not reply in next few hours, because here is 1:50 midnight and I am going to sleep.)

**Reply by longguzzz** [reply]
Final message before sleep ...
![image](https://user-images.githubusercontent.com/96779663/199798122-db28a119-fc4c-4412-bfb0-88a31a4a8580.png)
The path is a little different from "appdata/lunarvim/lvim".
Actually, the folder lunarvim is not the child of appdata. 
And the config is in another place.
![image](https://user-images.githubusercontent.com/96779663/199800366-ad2108f5-e6c6-4b26-aa9f-94d9b82a8a21.png)

By the way, serveral few minutes ago, I found [nvim-lsp-installer](https://github.com/williamboman/nvim-lsp-installer) and [mason.nvim](https://github.com/williamboman/mason.nvim) will install rust-analyzer in different folder.
The default rust-analyzer is installed by mason in `AppData\Roaming\nvim-data\mason\packages\rust-analyzer`, and it will show "cmd is executable: true".
The other rust-analyzer installed by nvim-lsp-installer will show "cmd is executable: false", and it is in `AppData\Local\nvim-data\...`. Here is no snapshot because I have uninstall the old version

**Reply by LostInTheLogs** [reply]
We don't use nvim-lsp-installer, its creator deprecated it in favor of mason. And run `git checkout master` to switch branches, rolling won't be updated anymore. Sorry for the wrong path, I don't know much about nvim on windows, please mention your os in the question in the future. Goodnight

**Reply by longguzzz** [reply]
OK, thank you. 
![image](https://user-images.githubusercontent.com/96779663/199857832-14a957a1-ff12-4a7d-853e-a4b4742bd665.png)
But don't worry, I will try it on wsl now, at least your answer help me stop trying it on windows. Will it be supportted for windows some day?

**Reply by longguzzz** [reply]
![image](https://user-images.githubusercontent.com/96779663/199860912-5254d421-ab63-45d1-bcee-96e345855df6.png)
Now, it works in WSL. And all the process only need less than 1/10 time than debugging in windows system.😂

**Reply by LostInTheLogs** [reply]
It should work in windows too, it may need more work to set up though.

**Reply by LostInTheLogs** [reply]
You can open an issue about the windows installer installing the rolling branch, I don't think that should happen

**Reply by longguzzz** [reply]
OK. https://github.com/LunarVim/LunarVim/issues/3413
